### PR TITLE
feat(reqresp): track peer rate-limit backoff in SelfRateLimiter

### LINKED ISSUE #8033: PeerDAS: rate limit when syncing
### Describe the bug

Sometimes we can not sync a peer up to head, the #1 error of "Batch download error" is rate limit:

```
...
2025-07-03 20:34:01.640 | Jul-03 13:34:01.639[sync]          verbose: Batch download error id=Head, startEpoch=1240, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
-- | -- | --
  |   | 2025-07-03 20:34:01.639 | Jul-03 13:34:01.639[sync]          verbose: Batch download error id=Head, startEpoch=1242, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
  |   | 2025-07-03 20:34:01.638 | Jul-03 13:34:01.637[sync]          verbose: Batch download error id=Head, startEpoch=1241, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
  |   | 2025-07-03 20:34:01.627 | Jul-03 13:34:01.627[sync]          verbose: Batch download error id=Head, startEpoch=1236, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
  |   | 2025-07-03 20:34:01.551 | Jul-03 13:34:01.551[sync]          verbose: Batch download error id=Head, startEpoch=1239, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
  |   | 2025-07-03 20:34:01.551 | Jul-03 13:34:01.551[sync]          verbose: Batch download error id=Head, startEpoch=1237, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
  |   | 2025-07-03 20:34:01.550 | Jul-03 13:34:01.550[sync]          verbose: Batch download error id=Head, startEpoch=1235, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
  |   | 2025-07-03 20:34:01.549 | Jul-03 13:34:01.549[sync]          verbose: Batch download error id=Head, startEpoch=1238, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
  |   | 2025-07-03 20:34:01.549 | Jul-03 13:34:01.548[sync]          verbose: Batch download error id=Head, startEpoch=1233, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited |  
  |   | 2025-07-03 20:34:01.548 | Jul-03 13:34:01.548[sync]          verbose: Batch download error id=Head, startEpoch=1232, status=Downloading, peer=16...2dKKZW, code=REQUEST_ERROR_INVALID_REQUEST, errorMessage=sNaPpYuV~2rate limited

...
```

Image

there are are logs like:

```
2025-07-03 20:34:22.740	
Jul-03 13:34:22.740[sync]          verbose: Batch download error id=Head, startEpoch=1199, status=Downloading, peer=16...bA5Tw4, code=REQUEST_ERROR_UNKNOWN_ERROR_STATUS, errorMessage=HsNaPpYLX:Rate limited. There are already 2 active requests with the same protocol, status=139
```

### Expected behavior

we should manage how many active requests in a peer
also it seems we download batches too aggressive, we should only download `n` epochs ahead 

### Steps to reproduce

_No response_

### Additional context

_No response_

### Operating system

Linux

### Lodestar version or commit hash

peerDAS

### PR DESCRIPTION
**Pull request by wemeetagain**
## Summary

Implements rate-limit backoff handling across the reqresp and sync layers, closing the loop on #8033.

When a peer rate-limits us (either via their response or our own `SelfRateLimiter`), we now:
1. **Don't count it as a failed download attempt** — batch reverts to `AwaitingDownload` via `downloadingRateLimited()` instead of incrementing the retry counter
2. **Back off the peer** — both the reqresp `SelfRateLimiter` and the sync `SyncChain` track the peer for a backoff period
3. **Fall back to other peers** — `ChainPeersBalancer` skips backed-off peers, so other available peers pick up the work immediately
4. **Parse explicit backoff durations** — Grandine's "Wait 2.816s" format is parsed and used directly; all other clients fall back to `DEFAULT_RATE_LIMIT_BACKOFF_MS` (5s)

This also restores `MAX_BATCH_DOWNLOAD_ATTEMPTS` from 20 back to 5 — the 20 was a workaround for the missing rate-limit handling.

### Cross-client rate limit behavior (as of March 2026)

| Client | Mechanism | Error message | Backoff hint |
|---|---|---|---|
| Lighthouse | Concurrent limit (max 2 in-flight/protocol) | `"Rate limited. There are already 2 active requests..."` | No |
| Prysm | 30s leaky bucket (blocks/blobs) | `"rate limited"` | No |
| Teku | 60s sliding window → disconnects | `"Peer has been rate limited"` | No |
| Nimbus | Silent token bucket throttling | Never sends errors | N/A |
| Grandine | GCRA with explicit wait | `"Wait 2.816488536s"` | Yes (parsed) |

### Changes by layer

**reqresp** (`packages/reqresp/`):
- `SelfRateLimiter`: added `rateLimitedUntilByPeer` map, `onRateLimited()`, `getRateLimitedPeerCount()`
- Backoff check at start of `allows()` with `lastSeenRequestsByPeer` update to prevent premature eviction
- `ReqResp.sendRequest()`: calls `onRateLimited()` when `RESP_RATE_LIMITED` is detected
- `responseStatusErrorToRequestError()`: improved cross-client documentation, reclassified Grandine "Wait X" from `RESP_TIMEOUT` to `RESP_RATE_LIMITED`
- `parseBackoffSeconds()`: parses Rust Duration format with hardened regex (bounded quantifiers, 256-char truncation)
- `RESP_RATE_LIMITED` type now carries optional `backoffMs`
- New metric: `beacon_reqresp_self_rate_limiter_rate_limited_peer_count`
- Constants: `DEFAULT_RATE_LIMIT_BACKOFF_MS = 5_000`, `MAX_RATE_LIMIT_BACKOFF_MS = 60_000`

**sync** (`packages/beacon-node/`):
- `SyncChain`: added `rateLimitedPeers` map, filters backed-off peers in `requestBatches()` before peer selection
- `sendBatch()`: handles `RESP_RATE_LIMITED` / `REQUEST_SELF_RATE_LIMITED` without counting as failed download
- `Batch`: added `downloadingRateLimited()` — reverts `Downloading` → `AwaitingDownload` without incrementing retry counter
- `unknownBlock.ts`: added `RESP_RATE_LIMITED` and `REQUEST_SELF_RATE_LIMITED` to no-exclude list
- `MAX_BATCH_DOWNLOAD_ATTEMPTS` restored from 20 to 5

## Test plan

- [x] **reqresp unit**: 7 tests for `SelfRateLimiter` backoff (blocking, expiry, custom duration, cap, keepalive, peer count, cleanup)
- [x] **reqresp unit**: 11 tests for `parseBackoffSeconds` (Grandine format, seconds variants, no-match cases, edge cases)
- [x] **reqresp e2e**: wire-level test — responder sends `RespStatus.RATE_LIMITED` (status 139) → requester detects `RESP_RATE_LIMITED` → second request blocked by `SelfRateLimiter` as `REQUEST_SELF_RATE_LIMITED`
- [x] **sync unit**: SyncChain with rate-limited peer — verifies batch retry counter is not exhausted, chain falls back to alternate peer and completes sync (TDD-verified: fails without feature, passes with it)
- [x] **existing tests**: all pass (batch unit tests, chain unit tests)

Closes #8033

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by wemeetagain** [comment]
@codex review
/gemini review

File: packages/beacon-node/src/sync/unknownBlock.ts [hunk 1]
```diff
@@ -569,6 +569,8 @@ export class BlockInputSync {
           downloadByRootMetrics?.error.inc({code: "req_resp", client: peerClient});
           switch (e.type.code) {
             case RequestErrorCode.REQUEST_RATE_LIMITED:
+            case RequestErrorCode.RESP_RATE_LIMITED:
+            case RequestErrorCode.REQUEST_SELF_RATE_LIMITED:
```
Comment (twoeths) [inline packages/beacon-node/src/sync/unknownBlock.ts:1244]: looks like this is not handled well as in range sync
when we got these errors, we should track it in a `private readonly rateLimitedPeers = new Map<PeerIdStr, number>();`
in order not to request them again immediately

**Comment by wemeetagain** [comment]
@lodekeeper please review

File: packages/reqresp/src/request/errors.ts [hunk 2]
```diff
@@ -98,6 +99,23 @@ export function responseStatusErrorToRequestError(e: ResponseError): RequestErro
...
+ * Returns the duration in milliseconds, or undefined if no parseable duration is found.
+ */
+export function parseBackoffSeconds(errorMessageLowercase: string): number | undefined {
+  // Match patterns like "2.816s", "30s", "2.816488536s", "30 seconds", "30 second"
+  const match = errorMessageLowercase.match(/(\d+(?:\.\d+)?)\s*(?:s(?:ec(?:ond)?s?)?)\b/);
+  if (match?.[1]) {
+    const seconds = Number.parseFloat(match[1]);
+    if (Number.isFinite(seconds) && seconds > 0) {
+      return Math.ceil(seconds * 1000);
+    }
```
[the remark this answers is not in the document]
Comment (lodekeeper) [reply packages/reqresp/src/request/errors.ts]: The effective cap is already enforced downstream in `SelfRateLimiter.onRateLimited()` via `Math.min(backoffMs, MAX_RATE_LIMIT_BACKOFF_MS)`, so a malicious peer cannot extend the actual backoff window past 60s. At this point the parser is only extracting a hint from the peer message; the enforcement point is the rate limiter.

File: packages/reqresp/test/unit/request/errors.test.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    {input: "wait 1 second", expected: 1000},
+    {input: "please wait 5 sec before retrying", expected: 5000},
+    // No parseable duration
+    {input: "rate limited", expected: undefined},
+    {input: "peer has been rate limited", expected: undefined},
+    {input: "rate limited. there are already 2 active requests with the same protocol", expected: undefined},
+    {input: "wait a moment", expected: undefined},
+    // Edge cases
+    {input: "wait 0s", expected: undefined},
+  ];
```
[the remark this answers is not in the document]
Comment (lodekeeper) [reply packages/reqresp/test/unit/request/errors.test.ts]: I don't think the parser test should assert the cap directly, because the cap is applied in `SelfRateLimiter.onRateLimited()`, not in `parseBackoffSeconds()`. The parser currently just extracts a numeric hint; the rate limiter is the component that clamps it to `MAX_RATE_LIMIT_BACKOFF_MS`.

File: packages/reqresp/src/rate_limiter/selfRateLimiter.ts [hunk 4]
```diff
@@ -57,6 +64,17 @@ export class SelfRateLimiter {
    */
   allows(peerId: PeerIdStr, protocolId: ProtocolID, requestId: RequestId): boolean {
     const now = Date.now();
+
+    // Check if peer has rate-limited us and we're still in backoff
+    const rateLimitedUntil = this.rateLimitedUntilByPeer.get(peerId);
+    if (rateLimitedUntil !== undefined) {
+      if (now < rateLimitedUntil) {
+        return false;
```
[the remark this answers is not in the document]
Comment (lodekeeper) [reply packages/reqresp/src/rate_limiter/selfRateLimiter.ts:77]: This one looks addressed in the current head — `allows()` now refreshes `lastSeenRequestsByPeer` before the early return in the backoff branch, so `checkDisconnectedPeers()` won't evict the backoff entry just because we kept attempting during the blocked window.

File: packages/reqresp/src/rate_limiter/selfRateLimiter.ts [hunk 5]
```diff
@@ -17,11 +17,18 @@ const DISCONNECTED_TIMEOUT_MS = 60 * 1000;
...
+export const MAX_RATE_LIMIT_BACKOFF_MS = 60_000;
+
 type RequestId = number;
 type RequestIdMs = number;
 
 /**
  * Simple rate limiter that allows a maximum of 2 concurrent requests per protocol per peer.
+ * Also tracks when a peer has rate-limited us and enforces a backoff period before allowing new requests.
  * The consumer should either prevent requests from being sent when the limit is reached or handle the case when the request is not allowed.
  */
```
[the remark this answers is not in the document]
Comment (lodekeeper) [reply packages/reqresp/src/rate_limiter/selfRateLimiter.ts:33]: I agree this would be a nice cleanup, but I wouldn't block this PR on it. The rest of the reqresp/sync timing code in this area is already using `Date.now()`, and switching just this class to a monotonic clock would be a broader consistency refactor. I'd keep the current approach here unless we want to do the clock-source cleanup more systematically.

File: packages/reqresp/src/rate_limiter/selfRateLimiter.ts [hunk 6]
```diff
@@ -95,16 +118,37 @@ export class SelfRateLimiter {
     trackedRequests.delete(requestId);
   }
 
+  /**
+   * Called when a peer responds with a rate-limit error, to enforce a backoff period before retrying.
+   */
+  onRateLimited(peerId: PeerIdStr, backoffMs = DEFAULT_RATE_LIMIT_BACKOFF_MS): void {
+    const cappedBackoffMs = Math.min(backoffMs, MAX_RATE_LIMIT_BACKOFF_MS);
+    const rateLimitedUntil = Date.now() + cappedBackoffMs;
+    this.rateLimitedUntilByPeer.set(peerId, rateLimitedUntil);
```
[the remark this answers is not in the document]
Comment (lodekeeper) [reply packages/reqresp/src/rate_limiter/selfRateLimiter.ts:126]: I agree with this one. `onRateLimited()` should preserve the longer existing backoff window instead of letting a shorter later response shrink it. If two in-flight requests to the same peer fail with different backoffs, we should keep the max-until timestamp.

File: packages/beacon-node/src/sync/unknownBlock.ts (hunk 1 above)
Comment (lodekeeper) [reply packages/beacon-node/src/sync/unknownBlock.ts:1244]: Agreed — the current `unknownBlock` change stops excluding the peer for rate-limit errors, but it still doesn't avoid immediately reselecting the same peer on the next attempt. I think this path wants the same kind of `rateLimitedPeers` backoff map that `SyncChain` now has, otherwise we keep hot-looping a peer that just told us to back off.

**Review by lodekeeper — COMMENTED** [review]
Looked through the latest patch — overall I think the direction is good and it closes the loop on the original sync/backoff failure mode in `#8033`.

My take on the current open review threads:
- **Already addressed in the latest head:**
  - `allows()` now refreshes `lastSeenRequestsByPeer` in the blocked/backoff branch, so the backoff entry won't get prematurely cleaned up
  - effective backoff capping already happens in `SelfRateLimiter.onRateLimited()` via `MAX_RATE_LIMIT_BACKOFF_MS`
- **I agree are still worth fixing:**
  - `onRateLimited()` should preserve the longest existing backoff window instead of letting a shorter later response shrink it
  - `unknownBlock` still wants a sync-layer `rateLimitedPeers` map like `SyncChain`, otherwise rate-limited peers can still get hot-looped immediately on re-selection
- **Reasonable but non-blocking cleanup:**
  - monotonic clock for `SelfRateLimiter` would be nice, but I'd treat that as a broader consistency cleanup rather than something to block this PR on by itself

So from my side the remaining substantive work is mainly the **max(existing backoff, new backoff)** behavior and the **unknownBlock peer backoff tracking**.

**Comment by lodekeeper** [comment]
Reviewed the latest head and left a detailed review. Main remaining items from my side are: (1) preserve the longest existing backoff window when multiple rate-limit responses hit the same peer, and (2) add unknownBlock-side peer backoff tracking similar to SyncChain so we don’t immediately reselect a peer that just rate-limited us.

File: packages/reqresp/src/request/errors.ts [hunk 7]
```diff
@@ -98,6 +107,36 @@ export function responseStatusErrorToRequestError(e: ResponseError): RequestErro
...
+ * produced by Rust's `format!("Wait {:?}", duration)`. All other clients send
+ * static messages with no backoff hint.
+ *
+ * Returns the duration in milliseconds, or undefined if no parseable duration is found
+ * (in which case the caller falls back to DEFAULT_RATE_LIMIT_BACKOFF_MS).
+ *
+ * Input comes from untrusted remote peer error messages — the regex uses bounded
+ * quantifiers and the input is truncated to defend against adversarial strings.
+ */
+export function parseBackoffSeconds(errorMessageLowercase: string): number | undefined {
```
Comment (nflaig) [inline packages/reqresp/src/request/errors.ts]: do we really need this? if it's only grandine it doesn't seem that useful to me, and removing it would simplify this PR

Comment (lodekeeper) [reply packages/reqresp/src/request/errors.ts]: Agreed — `parseBackoffSeconds` has narrow utility:

- Only Grandine produces parseable durations. Lighthouse switched to silent throttling (Apr 2025), Prysm/Teku send static text, Nimbus is silent.
- `DEFAULT_RATE_LIMIT_BACKOFF_MS` is 5s. Grandine's GCRA hint can be shorter or longer than that, so honoring it is at most a small throughput delta against one client.

If we drop the parser, we can also strip `backoffMs?: number` from the `RESP_RATE_LIMITED` type and the `backoffMs` plumbing through `SelfRateLimiter.onRateLimited` — that's where most of the surface area in this PR lives, and it removes the regex-on-untrusted-input concern entirely.

Worth keeping independently: the `wait ` branch reclassification (`RESP_TIMEOUT` → `RESP_RATE_LIMITED`). That's a correctness fix that makes Grandine actually trigger the self-rate-limit path. The branch becomes a one-liner without the parser.

My earlier comments on the cap (#discussion_r3000981163) and the parser test (#discussion_r3000981307) become moot if we go this route.

Comment (wemeetagain) [reply packages/reqresp/src/request/errors.ts]: yea delete

**Review by nflaig — APPROVED** [review]
LGTM

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.43.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.43.0) :tada:
